### PR TITLE
Make representation computation branchless in TransformerEncoderBase

### PR DESCRIPTION
**Pull request by zhxchen17**
Summary:
We want to make the computation branchless here because fairseq code may be exported and traced for deployment purposes, and tracing mechanisms can break the correctness for a captured program if it's dependent on input data. In this diff we try to rewrite the code to remove one branch so that tracer can proceed here and preserve the correct semantics of the model.

Test Plan:
CI

Reviewers:

Subscribers:

Tasks:

Tags:

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by zhxchen17** [comment]
cc @suo

**Comment by suo** [comment]
@dianaml0 do you mind taking a look at this?

**Comment by tugsbayasgalan** [comment]
cc @alexeib

File: fairseq/models/transformer/transformer_encoder.py [hunk 1]
```diff
@@ -207,8 +207,7 @@ def forward_scriptable(
         x, encoder_embedding = self.forward_embedding(src_tokens, token_embeddings)
 
         # account for padding while computing the representation
-        if has_pads:
-            x = x * (1 - encoder_padding_mask.unsqueeze(-1).type_as(x))
+        x = x * (1 - encoder_padding_mask.unsqueeze(-1).type_as(x) * has_pads.type_as(x))
```
Comment (moslehpour) [inline fairseq/models/transformer/transformer_encoder.py]: isn't `has_pads ` a boolean? 'bool' object doesn't have 'type_as' attribute!

Comment (zhxchen17) [reply fairseq/models/transformer/transformer_encoder.py]: has_pads is a `Tensor` object with a bool scalar, which could be converted by `type_as` method on `Tensor` type. That's my understanding.

Comment (zhxchen17) [reply fairseq/models/transformer/transformer_encoder.py]: actually, if the device is "xla" for src_token, it would be a bool which is rare but possible.

Let me update the PR to handle that.

Comment (zhxchen17) [reply fairseq/models/transformer/transformer_encoder.py]: @moslehpour updated. mind looking again?

Comment (moslehpour) [reply fairseq/models/transformer/transformer_encoder.py]: Great. This should work! 
Just one last comment, can you confirm if torchscript support type casting with `type_as`?

Comment (zhxchen17) [reply fairseq/models/transformer/transformer_encoder.py]: @moslehpour sure, I just tried that:
```
@torch.jit.script
def foo(x, y):
    return x.type_as(y)

a = foo(torch.tensor(True), torch.ones(3, 2))
print(a)
```
and seems Torchscript can give us correct result:
```
tensor(1.)
```

Comment (moslehpour) [reply fairseq/models/transformer/transformer_encoder.py]: Perfect. Thanks for checking it.

**Review by moslehpour — APPROVED** [review]

**Comment by alexeib** [comment]
thanks!

**Comment by alexeib** [comment]
the tests are failing now because of this change:

Variable 'has_pads' is annotated with type Tensor but is being assigned to a value of type bool:
  File "/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/fairseq/models/transformer/transformer_encoder.py", line 205
        # compute padding mask
        encoder_padding_mask = src_tokens.eq(self.padding_idx)
        has_pads: Tensor = (
        ~~~~~~~~ <--- HERE
            torch.tensor(src_tokens.device.type == "xla") or encoder_padding_mask.any()
        )
'TransformerEncoderBase.forward_scriptable' is being compiled since it was called from 'TransformerEncoderBase.forward'
  File "/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/fairseq/models/transformer/transformer_encoder.py", line 165
                  Only populated if *return_all_hiddens* is True.
        """
        return self.forward_scriptable(
               ~~~~~~~~~~~~~~~~~~~~~~~~
            src_tokens, src_lengths, return_all_hiddens, token_embeddings
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE

**Comment by zhxchen17** [comment]
> the tests are failing now because of this change:
> 
> Variable 'has_pads' is annotated with type Tensor but is being assigned to a value of type bool: File "/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/fairseq/models/transformer/transformer_encoder.py", line 205 # compute padding mask encoder_padding_mask = src_tokens.eq(self.padding_idx) has_pads: Tensor = ( ~~~~~~~~ <--- HERE torch.tensor(src_tokens.device.type == "xla") or encoder_padding_mask.any() ) 'TransformerEncoderBase.forward_scriptable' is being compiled since it was called from 'TransformerEncoderBase.forward' File "/opt/hostedtoolcache/Python/3.9.15/x64/lib/python3.9/site-packages/fairseq/models/transformer/transformer_encoder.py", line 165 Only populated if _return_all_hiddens_ is True. """ return self.forward_scriptable( ~~~~~~~~~~~~~~~~~~~~~~~~ src_tokens, src_lengths, return_all_hiddens, token_embeddings ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE

@alexeib I'm looking into this issue, and I'll try to send in a fix real quick.
